packages/xo-server/src/xo-mixins/proxies.js

```diff
@@ -132,79 +133,109 @@ export default class Proxy {
-
-      const [
-        password,
-        { registrationToken, registrationEmail: email },
-      ] = await Promise.all([generateToken(10), app.getApplianceRegistration()])
-      xenstoreData = {
-        'vm-data/system-account-xoa-password': password,
-        'vm-data/xo-proxy-authenticationToken': JSON.stringify(
-          proxyAuthenticationToken
-        ),
-        'vm-data/xoa-updater-credentials': JSON.stringify({
-          email,
-          registrationToken,
-        }),
-        'vm-data/xoa-updater-channel': JSON.stringify(xoProxyConf.channel),
-      }
-      if (networkConfiguration !== undefined) {
-        xenstoreData['vm-data/ip'] = networkConfiguration.ip
-        xenstoreData['vm-data/gateway'] = networkConfiguration.gateway
-        xenstoreData['vm-data/netmask'] = networkConfiguration.netmask
-        xenstoreData['vm-data/dns'] = networkConfiguration.dns
-      }
+    const props = { licenseId, boundObjectId: vm.uuid }
+    await app.bindLicense(props)
+    $defer.onFailure(() => app.unbindLicense(props))
+
+    if (networkId !== undefined) {
       await Promise.all([
-        vm.add_tags(xoProxyConf.vmTag),
-        vm.set_name_label(this._generateDefaultVmName(date)),
-        vm.update_xenstore_data(xenstoreData),
+        ...vm.VIFs.map(vif => xapi.deleteVif(vif)),
+        xapi.createVif(vm.$id, networkId),
       ])
+    }
 
-      await xapi.startVm(vm.$id)
-    } catch (error) {
-      await xapi._deleteVm(vm)
-      throw error
+    const date = new Date()
+    const proxyAuthenticationToken = await generateToken()
+
+    const [
+      password,
+      { registrationToken, registrationEmail: email },
+    ] = await Promise.all([generateToken(10), app.getApplianceRegistration()])
+    const xenstoreData = {
+      'vm-data/system-account-xoa-password': password,
+      'vm-data/xo-proxy-authenticationToken': JSON.stringify(
+        proxyAuthenticationToken
+      ),
+      'vm-data/xoa-updater-credentials': JSON.stringify({
+        email,
+        registrationToken,
+      }),
+      'vm-data/xoa-updater-channel': JSON.stringify(xoProxyConf.channel),
     }
+    if (networkConfiguration !== undefined) {
+      xenstoreData['vm-data/ip'] = networkConfiguration.ip
+      xenstoreData['vm-data/gateway'] = networkConfiguration.gateway
+      xenstoreData['vm-data/netmask'] = networkConfiguration.netmask
+      xenstoreData['vm-data/dns'] = networkConfiguration.dns
+    }
+    await Promise.all([
+      vm.add_tags(xoProxyConf.vmTag),
+      vm.set_name_label(this._generateDefaultVmName(date)),
+      vm.update_xenstore_data(xenstoreData),
+    ])
+
+    await xapi.startVm(vm.$id)
+
+    return {
+      date,
+      proxyAuthenticationToken,
+      vm,
+      xenstoreData,
+    }
+  }
+
+  async deployProxy(
+    srId,
+    licenseId,
+    { networkConfiguration, networkId, proxyId } = {}
+  ) {
+    const app = this._app
+
+    const redeploy = proxyId !== undefined
+    if (redeploy) {
```

Comment:
Should we throw if both a `licenseId` and a `proxyId` have been provided?

Reply:
No need, the redeploy will be implemented in the next PR and the main PR will not be merged until all cases are handled.

Reply:
So the redeploy will accept a `licenseId`? And if so, why?

Reply:
A proxy VM can be not bond to a license that's why `licenceId` is required in the server part.

Reply:
Ok, and the proxy VM has to be redeployed for this? We can't bind a license without deploying?